include/mata/sparse-set.hh

```diff
@@ -0,0 +1,355 @@
+        static_assert(std::is_unsigned<Number>::value, "SparseSet can only contain unsigned integers");
+
+    private:
+        std::vector<Number> dense;    //Dense set of elements
+        std::vector<Number> sparse;    //Map of elements to dense set indices
+
+        size_t size_ = 0;    //Current size (number of elements)
+        size_t capacity_ = 0;    //Current capacity (maximum value + 1 or more (can be more when maximum is removed))
+
+    public:
+        using iterator = typename std::vector<Number>::const_iterator;
+        using const_iterator = typename std::vector<Number>::const_iterator;
+
+        iterator begin() { return dense.begin(); }
+
+        const_iterator begin() const { return dense.begin(); }
+
+        iterator end() { return dense.begin() + size_; }
+
+        const_iterator end() const { return dense.begin() + size_; }
+
+        size_t size() const { return size_; }
+
+        size_t capacity() const { return capacity_; }
+
+        bool empty() const { return size_ == 0; }
+
+        void clear() { size_ = 0;  }
+
+        // TODO: maybe we could reserve space more efficiently, by something as doubling?
+        //  But we should not create havoc with capacity, which is used outside, namely for determining the states of an automaton.
+        void reserve(size_t u) {
+            if (u > capacity_) {
+                dense.resize(u, 0);
+                sparse.resize(u, 0);
+                capacity_ = u;
+            }
+
+            assert(consistent());
+        }
+
+        bool has(const Number val) const {
+            return val < capacity_ &&
+                   sparse[val] < size_ &&
+                   dense[sparse[val]] == val;
+        }
+
+        void insert(const Number val) {
+            assert(consistent());
+
+            if (!has(val)) {
+                if (val >= capacity_)
+                    reserve(val + 1);
+
+                dense[size_] = val;
+                sparse[val] = size_;
+                ++size_;
+            }
+
+            assert(consistent());
+        }
+
+        void erase(const Number val) {
+            if (has(val)) {
+                dense[sparse[val]] = dense[size_ - 1];
+                sparse[dense[size_ - 1]] = sparse[val];
+                --size_;
+            }
+        }
+
+
+
+//////////////////////////////////////////////////////////////////////////////////
+///     New Mata code
+//////////////////////////////////////////////////////////////////////////////////
+
+// Constructors:
+
+        SparseSet(Number size = 0, bool val = false): sparse(size, val), dense(size) {
+            if (val) {
+                size_ = size;
+                for (Number e = 0; e < size; ++e)
+                    dense.push_back(e);
+            }
+            else
+                size_ = 0;
+
+            assert(consistent());
+        };
+
+        //TODO: should this be explicit too? it would mean a lot of fixing expecially in tests, unfortunatelly
```

Comment:
No, it should not. Remove the comment.

Reply:
Ok, but could you add explanation? Here or elsewhere? :)